### PR TITLE
Instantiation expression can be followed by line break or binary operator

### LINKED ISSUE #49348: Syntax Error for `let` var declaration under semicolon-less instantiation expressions
# Bug Report

### 🔎 Search Terms

- `instantiation expression semicolon`

### 🕗 Version & Regression Information

- This is a bug of new syntax that is landed in TS 4.7

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/MYewdgzgLgBAZiEMC88CuZgB5oCcCWYA5gHwBQZANgKawBGAhrgNxlA)

### 💻 Code

```ts
const foo = func<string>

let bar;
```

### 🙁 Actual behavior

Syntax error `Cannot find name 'let'.(2304)`

### 🙂 Expected behavior

Parsing successfully

### LINKED ISSUE #49362: Instantiation expressions not allowed in logical/coalescing expressions
```ts
declare let f: undefined | (<T>(x: T) => T);

let g = f<string> || ((x: string) => string);
```

> **Maybe OT** Since you are fine-tuning how instantiation expressions are parsed: it's weird that `a<b> ?? c` is not valid code, from a syntactic point of view.

_Originally posted by @nicolo-ribaudo in https://github.com/microsoft/TypeScript/pull/49353#issuecomment-1145392217_

### PR DESCRIPTION
**Pull request by ahejlsberg**
This PR adjusts our parsing of instantiation expressions. When a potential type argument list is followed by

* a line break,
* an `(` token,
* a template literal string, or
* any token except `<` or `>` that isn't the start of an expression,

we consider that construct to be a type argument list. Otherwise we consider the construct to be a `<` relational expression followed by a '>` relational expression.

Fixes #49348.
Fixes #49362.

### CODE REVIEW

File: src/compiler/parser.ts [hunk 1]
```diff
@@ -5732,6 +5732,12 @@ namespace ts {
                 case SyntaxKind.NoSubstitutionTemplateLiteral:  // foo<T> `...`
                 case SyntaxKind.TemplateHead:                   // foo<T> `...${100}...`
                     return true;
+                // These strict mode reserved words are valid identifiers in non-strict mode, and can thus
+                // start an expression. However, when they follow a type argument list and are immediately
+                // preceded by a line break, we don't consider them expression starters.
```
Comment (DanielRosenwasser) [inline src/compiler/parser.ts:5737]:
```suggestion
                // Technically strict mode reserved words are valid identifiers in non-strict mode, and can thus
                // start an expression. However, when they follow a type argument list and are immediately
                // preceded by a line break, we don't consider them expression starters.
                // This supports a semicolon-less style for code like the following:
                //
                //   let specialFoo = foo<string>
                //   let value = ...
                //
```

File: src/compiler/parser.ts [hunk 2]
```diff
@@ -5732,6 +5732,17 @@ namespace ts {
...
+                // Technically strict mode reserved words are valid identifiers in non-strict mode, and can thus
+                // start an expression. However, when they follow a type argument list and are immediately
+                // preceded by a line break, we don't consider them expression starters.
+                // This supports a semicolon-less style for code like the following:
+                //
+                //   let specialFoo = foo<string>
+                //   let value = ...
+                //
+                case SyntaxKind.InterfaceKeyword:
+                case SyntaxKind.LetKeyword:
```
Comment (DanielRosenwasser) [inline src/compiler/parser.ts:5744]: You need to cover more than just these cases and have tests for them - for example

```ts
class C {
    static specialFoo = f<string>
    static bar = 123;
}

class D {
    public specialFoo = f<string>
    public bar = 123;
}

class E {
    private specialFoo = f<string>
    private bar = 123;
}

class F {
    protected specialFoo = f<string>
    protected bar = 123;
}
```

**Review by DanielRosenwasser — CHANGES_REQUESTED** [review]

**Comment by DanielRosenwasser** [comment]
CC @nicolo-ribaudo @evanw @kdy1 the idea of this PR and issue is that an instantiation expression cannot be followed by a newline and some set of strict-mode reserved words.

**Comment by nicolo-ribaudo** [comment]
Other than strict-mode keywords, you might also want to special-case `await` (which is only a keyword in modules, and not in general in strict mode).

**Comment by DanielRosenwasser** [comment]
We just discussed the nuances of `async`. We're considering a rule where if something that could be parsed as an instantiation expression is followed by an identifier or keyword of any sort, then we will favor the instantiation expression (and effectively act as if a user had added a semicolon).

**Comment by nicolo-ribaudo** [comment]
I hate any new "valid JavaScript has different semantics when parsed as TypeScript" rule, but it's so uncommon to see `a < b > [LineTerminator] c` in JS code that probably no one will never notice it.

**Comment by DanielRosenwasser** [comment]
> **Maybe OT** Since you are fine-tuning how instantiation expressions are parsed: it's weird that `a<b> ?? c` is not valid code, from a syntactic point of view.

😬

Opened https://github.com/microsoft/TypeScript/issues/49362

**Comment by sandersn** [comment]
@typescript-bot test top100

**Comment by sandersn** [comment]
Ran out memory trying to compiler angular ☠️ 
I'm going to try again, especially since I thought angular was part of the user tests too. Maybe it's an old version though.

@typescript-bot test top100

**Comment by sandersn** [comment]
Let's try again:
@typescript-bot test top100

**Comment by ahejlsberg** [comment]
@typescript-bot test this
@typescript-bot user test this inline
@typescript-bot run dt
@typescript-bot perf test faster

**Comment by ahejlsberg** [comment]
All test runs are clean and performance is unaffected.

**Review by RyanCavanaugh — APPROVED** [review]

**Comment by nicolo-ribaudo** [comment]
Does this PR also change the behavior of new expressions? e.g.
```ts
new A<B>
C
```
does ASI affect this code?

(or as an alternative, it there a playground where I can test it myself? :sweat_smile:)

**Comment by sosukesuzuki** [comment]
Hi, will this PR be released in the near future?

**Pull request merged**